### PR TITLE
Feature/logger

### PR DESCRIPTION
**Pull request by hanhanW**
* Implement logger
* Log the message when recv/send in the tube.

---
This change is [Reviewable](https://reviewable.io/reviews/peter50216/pwntools-ruby/40)

### CODE REVIEW

File: lib/pwnlib/context.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # encoding: ASCII-8BIT
 
+# (XXX): Avoid circular require
```
Comment (david942j) [inline lib/pwnlib/context.rb]: Why circular require happens?
Shouldn't the next line just load `::Logger` ?

File: test/tubes/tube_test.rb [hunk 2]
```diff
@@ -1,12 +1,38 @@
...
 require 'pwnlib/tubes/tube'
 
 class TubeTest < MiniTest::Test
-  include ::Pwnlib::Tubes
   include ::Pwnlib::Context
+  include ::Pwnlib::Tubes
+
+  def setup
+    # Default to disable coloring for easier testing.
+    Rainbow.enabled = false
```
Comment (david942j) [inline test/tubes/tube_test.rb]: set back to original value in `teardown`

**Comment by hanhanW** [comment]
Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[lib/pwnlib/context.rb, line 3 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KshUtOzwYDY_WwO1JRv-r1-3:-KsiC3DVqmP65MCVgbfd:b2nhexc) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/c34197cd073281acea88333dc110e1d10eb98ef7/lib/pwnlib/context.rb#L3)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

Why circular require happens?
Shouldn't the next line just load `::Logger` ?
</blockquote></details>

Fix

---

*[test/tubes/tube_test.rb, line 19 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KshUtOzwYDY_WwO1JS--r1-19:-KsiCXTZAQTWipVixzmk:bwrf6ap) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/c34197cd073281acea88333dc110e1d10eb98ef7/test/tubes/tube_test.rb#L19)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

set back to original value in `teardown`
</blockquote></details>

Disable in test_helper.rb

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40)*

**Comment by david942j** [comment]
Reviewed 2 of 6 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[lib/pwnlib/logger.rb, line 9 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KsibNngy7RHYql7yiL0:-KsibNngy7RHYql7yiL1:bbc8lm2) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L9)):*
> ```Ruby
> 
> module Pwnlib
>   # Logger module!
> ```

This comment is too meaningless XDD

---

*[lib/pwnlib/logger.rb, line 12 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-Ksibq1xDOe8Ci9hRkCh:-Ksibq1xDOe8Ci9hRkCi:b7imzsa) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L12)):*
> ```Ruby
>   module Logger
>     # The type for logger which inherits Ruby builtin Logger.
>     # We use +context.log_level+ instead of +@level+.
> ```

Main difference is using +context.log_level+ instead of +level+ in logging methods

---

*[lib/pwnlib/logger.rb, line 26 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-Ksibca7XfdsXFKz07oV:-Ksibca7XfdsXFKz07oW:b-xus2v4) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L26)):*
> ```Ruby
>         super(STDOUT)
>         # Set +@level+ to the lowest priority.
>         @level = DEBUG
> ```

`@level` is unnecessary?

---

*[lib/pwnlib/logger.rb, line 60 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KsicSPnaxm7rmnB7Wzz:-KsicSPnaxm7rmnB7X--:b-ldjn27) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L60)):*
> ```Ruby
> 
>     # @!attribute [r] logger
>     #   @return [LoggerType] the singleton logger for all class.
> ```

classes

---

*[lib/pwnlib/logger.rb, line 65 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-Ksick-ghEjscumFVdRY:-Ksick-ghEjscumFVdRZ:b-htfg5t) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L65)):*
> ```Ruby
>     end
> 
>     # A module for include hook for logger.
> ```

Include this module to use logger.

---

*[lib/pwnlib/logger.rb, line 66 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KsicbyQbH9PRdqzT2ck:-KsicbyQbH9PRdqzT2cl:b-1r83lb) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L66)):*
> ```Ruby
> 
>     # A module for include hook for logger.
>     # Including Pwnlib::Logger from module M would add +log+ as a private instance method and a private class
> ```

`{Pwnlib::Logger}`

---

*[test/tubes/tube_test.rb, line 19 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KshUtOzwYDY_WwO1JS--r1-19:-Ksie2Yz7fwGzMQ8sBqw:bco6ms9) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/c34197cd073281acea88333dc110e1d10eb98ef7/test/tubes/tube_test.rb#L19)):*
<details><summary><i>Previously, hanhanW wrote…</i></summary><blockquote>

Disable in test_helper.rb
</blockquote></details>

forget it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40)*

**Comment by hanhanW** [comment]
Review status: 7 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[lib/pwnlib/logger.rb, line 9 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KsibNngy7RHYql7yiL0:-KsiuRuw1__kkYe20T2Q:b-cq3lpf) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L9)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

This comment is too meaningless XDD
</blockquote></details>

It's more meaningful now

---

*[lib/pwnlib/logger.rb, line 12 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-Ksibq1xDOe8Ci9hRkCh:-Ksit3oRxsOmQmp9AuUi:byfjvzd) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L12)):*
> Main difference is using +context.log_level+ instead of +level+ in logging methods

Fix

---

*[lib/pwnlib/logger.rb, line 26 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-Ksibca7XfdsXFKz07oV:-Ksit5sPOIJn9W1lR_Om:b-elz0dz) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L26)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`@level` is unnecessary?
</blockquote></details>

Yes, you're right.
It's unnecessary since the default value of `@level` is DEBUG.

---

*[lib/pwnlib/logger.rb, line 60 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KsicSPnaxm7rmnB7Wzz:-KsitkrA-mQEjR0lzCy-:b2nhexc) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L60)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

classes
</blockquote></details>

Fix

---

*[lib/pwnlib/logger.rb, line 65 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-Ksick-ghEjscumFVdRY:-KsitroaXn0ftch3Rbt-:bv3s70) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L65)):*
> Include this module to use logger.
Fix

---

*[lib/pwnlib/logger.rb, line 66 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KsicbyQbH9PRdqzT2ck:-KsitsVOWTdBWoJHmF02:b2nhexc) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/0798b983c9ab8c3316c2017412809c73908ffae6/lib/pwnlib/logger.rb#L66)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`{Pwnlib::Logger}`
</blockquote></details>

Fix

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40)*

File: lib/pwnlib/logger.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# encoding: ASCII-8BIT
+
+require 'logger'
+require 'rainbow'
+
+require 'pwnlib/context'
+
+module Pwnlib
+  # Logging module for printing status during an exploit, and internally within {Pwnlib}.
```
Comment (david942j) [inline lib/pwnlib/logger.rb]: exploitation

File: lib/pwnlib/logger.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  #
+  # An exploit developer can use +log+ to print out status messages during an exploitation.
+  module Logger
+    # The type for logger which inherits Ruby builtin Logger.
+    # Main difference is using +context.log_level+ instead of +level+ in logging methods
```
Comment (david942j) [inline lib/pwnlib/logger.rb]: missing `.` in the end

**Comment by hanhanW** [comment]
Review status: 6 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[lib/pwnlib/logger.rb, line 9 at r3](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KshUtOzwYDY_WwO1JRw-r3-9:-KtB0IQ6mgX4QotfkMnR:b2ni2nk) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/38bf4615d6e2159bb8453206c98e6df91ae32efe/lib/pwnlib/logger.rb#L9)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

exploitation
</blockquote></details>

fix

---

*[lib/pwnlib/logger.rb, line 14 at r3](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-KshUtOzwYDY_WwO1JRw-r3-14:-KtB0JUsAnp6DDPtYocJ:b2ni2nk) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/38bf4615d6e2159bb8453206c98e6df91ae32efe/lib/pwnlib/logger.rb#L14)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

missing `.` in the end
</blockquote></details>

fix

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40)*

**Comment by david942j** [comment]
:+1:
:lgtm: 

---

Reviewed 2 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/40#-:-KtBIwbuJf--mbDStTxm:bkxnm95)*

**Pull request merged**